File: changuito/Assets/Scripts/Architecture/Persistence/SharedObject.cs
using UnityEngine;
using System.Collections;
using System.Text;
using System.Collections.Generic;
using System;

public class SharedObject
{
    private Dictionary<string, object> _data;
    
    public SharedObject()
    {
        _data = new Dictionary<string, object>(20);
    }
    
    private SharedObject(Dictionary<string, object> data)
    {
        _data = data;
    }

    public static SharedObject Deserialize(byte[] bytes)
    {
        return Deserialize(Encoding.UTF8.GetString(bytes));
    }

    public static SharedObject Deserialize(string json)
    {
        return new SharedObject((Dictionary<string,object>)MiniJSON.Json.Deserialize(json));
    }
    
    public static T CastSharedObject<T>(SharedObject parent, string key, SharedObject child)
        where T: SharedObject
    {
        T childAsT = child as T;
        if (child != null && childAsT == null)
        {
            childAsT = (T)Activator.CreateInstance(typeof(T));
            childAsT.MergeWith(child);
            parent.Set(key, childAsT);
        }
        return childAsT;
    }

    public override string ToString()
    {
        return MiniJSON.Json.Serialize(_data);
    }

    public void AddToList<T>(string key, List<T> list)
    {
        list.ForEach(element => AddToList(key, element));
    }
    
    public void AddToList<T>(string key, T value)
    {
        Set(key + (GetList<T>(key).Count + 1), value);
    }

    public List<T> GetSharedObjectList<T>(string key, Func<string, SharedObject, T> converter)
        where T: SharedObject
    {
        return GetList<T>(key, delegate(string childKey)
        {
            SharedObject sharedObject = GetSharedObject(childKey);
            if (sharedObject == null)
                return  null;
            return converter(childKey, sharedObject);
        });
    }

    public List<T> GetSharedObjectList<T>(string key)
        where T: SharedObject
    {
        return GetList<T>(key, (string childKey) => GetSharedObject<T>(childKey));
    }
    
    public List<T> GetList<T>(string key)
    {
        return GetList<T>(key, (string childKey) => GetValueOrDefault<T>(childKey));
    }

    public bool GetBool(string key)
    {
        return GetValueOrDefault<bool>(key);
    }

    public double GetDouble(string key)
    {
        return GetValueOrDefault<double>(key);
    }

    public int GetInt(string key)
    {
        return GetValueOrDefault<int>(key);
    }

    public long GetLong(string key)
    {
        return GetValueOrDefault<long>(key);
    }
    
    public T GetSharedObject<T>(string key)
        where T: SharedObject
    {
        return CastSharedObject<T>(this, key, GetSharedObject(key));
    }
    
    public SharedObject GetSharedObject(string key)
    {
        object value = GetValueOrDefault<object>(key);
        
        SharedObject valueAsSharedObject = value as SharedObject;
        if (value != null && valueAsSharedObject == null)
        {
            // Preferible dejar el objeto ya deserializado, en vez de deserializarlo siempre
            valueAsSharedObject = SharedObject.Deserialize((string)value);
            Set(key, valueAsSharedObject);
        }
        return valueAsSharedObject;
    }
    
    public string GetString(string key)
    {
        return GetValueOrDefault<string>(key);
    }
    
    public void MergeWith(SharedObject sharedObject)
    { 
        foreach (var entry in sharedObject._data)
        {
            string key = entry.Key;
            object value = entry.Value;
            
            SharedObject valueAsSharedObject = value as SharedObject;
            if (valueAsSharedObject != null)
            {
                SharedObject oldValue = GetSharedObject(key);
                
                if (oldValue != null)
                {
                    oldValue.MergeWith(valueAsSharedObject);
                } else
                {
                    Set(key, valueAsSharedObject);
                }
            } else
            {
                Set(key, value);
            }
        }
    }
        
    public byte[] Serialize()
    {
        return Encoding.UTF8.GetBytes(ToString());
    }

    public void Set<T>(string key, T value)
    {
        _data [key] = value;
    }

    private bool ExistsKey(string key)
    {
        object value;
        return _data.TryGetValue(key, out value);
    }

    private T GetValueOrDefault<T>(string key)
    {
        object value = null;
        if (!_data.TryGetValue(key, out value))
            return default(T);
        if (value is T)
            return (T)value;
        return (T) Convert.ChangeType(value, typeof(T));
    }
    
    private List<T> GetList<T>(string key, Func<string, T> getItemDelegate)
    {
        List<T> items = new List<T>(15);
        
        int i = 1;
        T item = default(T);
        while (ExistsKey(key + i))
        {
            items.Add(getItemDelegate(key + i));
            i++;
        }
        return items;
    }
}

File: webserver/webserver/changuito/changuito-war/src/main/java/ar/com/utn/changuito/architecture/net/SharedObject.java
package ar.com.utn.changuito.architecture.net;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class SharedObject {

    private final Map<String, Object> data;

    public SharedObject() {
        data = new HashMap<String, Object>(20);
    }

    public SharedObject(final SharedObject sharedObject) {
        this(sharedObject.data);
    }

    private SharedObject(final Map<String, Object> data) {
        this.data = data;
    }

    public static SharedObject deserialize(final byte[] jsonCodeBytes) {

        return new SharedObject(JsonUtils.convertJsonBytesToMap(jsonCodeBytes));
    }

    public static SharedObject deserialize(final String jsonCode) {

        return new SharedObject(JsonUtils.convertJsonStringToMap(jsonCode));
    }

    public Set<String> getKeys() {
        return data.keySet();
    }

    public Object get(final String key) {
        final Object value = data.get(key);
        if (value instanceof SharedObject) {
            return value.toString();
        }
        return value;
    }

    public double getDouble(final String key) {
        return (Double) data.get(key);
    }

    public int getInt(final String key) {
        return (Integer) data.get(key);
    }

    public long getLong(final String key) {
        final Object value = data.get(key);
        if (value instanceof Integer) {
            return (long) (Integer) value;
        }
        return (Long) value;
    }

    public String getString(final String key) {
        return (String) data.get(key);
    }

    public <T extends SharedObject> T getSharedObject(final String key, final Class<T> classType) {

        try {
            final T result = classType.newInstance();
            result.mergeWith(getSharedObject(key));
            return result;
        } catch (final Exception exception) {
            throw new RuntimeException(exception.getMessage(), exception);
        }
    }

    public SharedObject getSharedObject(final String key) {

        final Object value = data.get(key);

        SharedObject valueAsSharedObject = null;
        if (value instanceof SharedObject) {
            valueAsSharedObject = (SharedObject) value;
        } else if (valueAsSharedObject == null) {
            valueAsSharedObject = deserialize((String) value);
            // Preferible dejar el objeto ya deserializado, en vez de deserializarlo siempre
            set(key, valueAsSharedObject);
        }
        return valueAsSharedObject;
    }

    public void mergeWith(final SharedObject sharedObject) {
        for (final Map.Entry<String, Object> entry : sharedObject.data.entrySet()) {
            final String key = entry.getKey();
            final Object value = entry.getValue();

            if (value instanceof SharedObject) {
                final SharedObject valueAsSharedObject = (SharedObject) value;
                final Object oldValue = getSharedObject(key);

                if (oldValue instanceof SharedObject) {
                    ((SharedObject) oldValue).mergeWith(valueAsSharedObject);
                } else {
                    set(key, valueAsSharedObject);
                }
            } else {
                set(key, value);
            }
        }
    }

    public byte[] serialize() {

        return JsonUtils.convertMapToJsonBytes(data);
    }

    public <T> void set(final String key, final T value) {
        if (value == null) {
            data.remove(key);
        } else {
            data.put(key, value);
        }
    }

    @Override
    public String toString() {

        return JsonUtils.convertMapToJsonString(data);
    }
}
